katdal/vis_flags_weights.py

```diff
@@ -279,18 +307,17 @@ def __init__(self, store, chunk_info, corrprods):
 
         vis = darray['correlator_data']
         # Combine low-resolution weights and high-resolution weights_channel
-        weights = darray['weights'] * darray['weights_channel'][..., np.newaxis]
-        # Scale weights according to power
+        raw_weights = darray['weights'] * darray['weights_channel'][..., np.newaxis]
+        # Scale weights according to power (or remove scaling if already applied)
         if corrprods is not None:
-            assert len(corrprods) == vis.shape[2]
-            # Ensure that we have only a single chunk on the baseline axis.
-            if len(vis.chunks[2]) > 1:
-                vis = vis.rechunk({2: vis.shape[2]})
-            if len(weights.chunks[2]) > 1:
-                weights = weights.rechunk({2: weights.shape[2]})
-            auto_indices, index1, index2 = corrprod_to_autocorr(corrprods)
-            weights = da.blockwise(weight_power_scale, 'ijk', vis, 'ijk', weights, 'ijk',
-                                   dtype=np.float32,
-                                   auto_indices=auto_indices, index1=index1, index2=index2)
-
-        VisFlagsWeights.__init__(self, vis, flags, weights, self.vis_prefix)
+            if raw_weights_are_unscaled:
+                weights = _scale_weights(vis, raw_weights, corrprods, divide=True)
+                unscaled_weights = raw_weights
+            else:
+                weights = raw_weights
+                unscaled_weights = _scale_weights(vis, raw_weights, corrprods, divide=False)
+        else:
+            # Don't bother with unscaled weights (it's optional)
+            weights = raw_weights
```

Comment:
Should it be an error if `raw_weights_are_unscaled` is true here?

Reply:
I like that - will add a ValueError.